Jetty 12 : precompressed content support for `ResourceService`

joakime opened the pull request:
Reenable test cases for precompressed content.

Fix implementation to support precompressed content.

joakime committed:
Remove Resource.getWeakETag in favor of new EtagUtil
Protect bad usages of CachingContentFactory

joakime committed:
Working precompressed content for ResourceService

joakime committed:
Fixing DefaultServlet handling of ResourceService.writeHttpError() overrides

joakime committed:
Protect NPE in CachingContentFactory.isValid()

joakime committed:
Complete callback in DefaultServlet.writeHttpError

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EtagUtil.java:
```diff
@@ -0,0 +1,194 @@
+     * @param suffix the suffix for the ETag
+     * @return the weak etag
+     */
+    public static String calcWeakEtag(Path path, String suffix)
```
Shouldn't this method accept a `Resource` instead of a `Path`?

joakime replied:
The weak etag is calculated from the path name + last modified timestamp.

Not all resources have them (eg: MemoryResource has neither).
Can't use a ResourceCollection here either, as the path name and last modified timestamp won't always represent the same path object.
It's better to use Path, as even with ResourceCollection in the mix, we'll get a Path then use it.
This is the same Path that is used by the HttpContent to serve the same content.

BTW, Don't get hung up on MemoryResource, the HttpContentFactory reorg PR coming after this will make the case to remove MemoryResource entirely.

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EtagUtil.java:
```diff
@@ -0,0 +1,194 @@
+     * @param path the path to calculate from
+     * @return the weak etag
+     */
+    public static String calcWeakEtag(Path path)
```
Shouldn't this method accept a `Resource` instead of a `Path`?

joakime replied:
Different method supports `Resource` now, this method still exists.

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ResourceHttpContent.java:
```diff
@@ -119,13 +119,13 @@ public HttpField getETag()
     public String getETagValue()
     {
-        return _resource.getWeakETag();
+        return EtagUtil.calcWeakEtag(_resource.getPath());
```
This would only work with path-backed resources. Shouldn't we be capable of calculating the etag of a potential `MemoryResource`?

joakime replied:
Fixed in this PR.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java:
```diff
@@ -405,7 +406,7 @@ public class CachedHttpContent implements HttpContent
             _lastAccessed = Instant.now();
 
-            _etag = CachedContentFactory.this._etags ? new PreEncodedHttpField(HttpHeader.ETAG, resource.getWeakETag()) : null;
+            _etag = CachedContentFactory.this._etags ? new PreEncodedHttpField(HttpHeader.ETAG, EtagUtil.calcWeakEtag(resource.getPath())) : null;
```
This would only work with path-backed resources. Shouldn't we be capable of calculating the etag of a potential `MemoryResource`?

joakime replied:
There's no ETag support for MemoryResource as it exists currently.

The HttpContentFactory reorg PR (after this one) will remove MemoryResource in favor of a StaticHttpContent object fallback instead (which can hold the etag based on the URI and JAR file last timestamp)

joakime replied:
Fixed in this PR anyway.

lorban commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ResourceHandlerTest.java:
```diff
@@ -750,8 +749,35 @@ public void testBigger() throws Exception
+        assertThat(response, containsHeaderValue(HttpHeader.VARY, "Accept-Encoding"));
+        assertThat(response, containsHeaderValue(HttpHeader.CONTENT_ENCODING, "br"));
+        String body = response.getContent();
+        assertThat(body, containsString("fake br"));
```
Why only `fake br`?

joakime replied:
The contents don't actually have to be compressed to satisfy the test.
The ResourceService doesn't compress it anyway, it just serves the content as-is from the Resource Base.

sbordet replied:
But it confuses the reader the fact that you test for `fake br` only, while at the beginning of the test the whole content is `fake brotli`. 
I would test for the whole content, since it's so small, to avoid such scratching head moment from other readers of this test ("why only `fake br`"?).

lorban commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ResourceHandlerTest.java:
```diff
@@ -1892,35 +1912,6 @@ public void testDirectoryOfCollections() throws Exception
     }
 
-    @Test
-    public void testEtagIfMatchAlwaysFailsDueToWeakEtag() throws Exception
```
Why did you remove this test?

joakime replied:
It's a bad/duplicate test.

There are other tests that test the same condition.
testIfEtag() for example.
And those expect a 200 OK, not a 412 precondition failure for the same request.

The origin of this test is unknown.
It is not present in jetty-11.0.x.
I can trace as far back as https://github.com/eclipse/jetty.project/commit/04acdb72f041a15267bb79c36d8fc4d517beaadb but not before then.
It seems to have been created during the hackathon.
But etags and preconditions were not working back then.

Interestingly, the github search also shows nothing for this test case...
https://github.com/eclipse/jetty.project/search?q=testEtagIfMatchAlwaysFailsDueToWeakEtag&type=code

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/PrecompressedHttpContent.java:
```diff
@@ -51,20 +51,19 @@ public HttpField getETag()
     {
-        //return _content.getResource().getWeakETag(_format.getEtagSuffix());
-        return null;
+        return EtagUtil.calcWeakEtag(_content.getResource().getPath(), _format.getEtagSuffix());
```
This would only work with path-backed resources. Shouldn't we be capable of calculating the etag of a potential `MemoryResource`?

joakime replied:
Fixing etag support

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/CachedContentFactory.java:
```diff
@@ -403,7 +404,7 @@ public class CachedHttpContent implements HttpContent
             _lastAccessed = Instant.now();
 
-            _etag = CachedContentFactory.this._etags ? new PreEncodedHttpField(HttpHeader.ETAG, resource.getWeakETag()) : null;
+            _etag = CachedContentFactory.this._etags ? new PreEncodedHttpField(HttpHeader.ETAG, EtagUtil.calcWeakEtag(resource.getPath())) : null;
```
This would only work with path-backed resources. Shouldn't we be capable of calculating the etag of a potential `MemoryResource`?

joakime replied:
Fixing.

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/CachedContentFactory.java:
```diff
@@ -603,7 +604,7 @@ public class CachedPrecompressedHttpContent extends PrecompressedHttpContent
             _precompressedContent = precompressedContent;
 
-            _etag = (CachedContentFactory.this._etags) ? new PreEncodedHttpField(HttpHeader.ETAG, _content.getResource().getWeakETag(format.getEtagSuffix())) : null;
+            _etag = (CachedContentFactory.this._etags) ? new PreEncodedHttpField(HttpHeader.ETAG, EtagUtil.calcWeakEtag(_content.getResource().getPath(), format.getEtagSuffix())) : null;
```
This would only work with path-backed resources. Shouldn't we be capable of calculating the etag of a potential `MemoryResource`?

joakime replied:
Fixing.

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-precompressed-support-resource-service

joakime committed:
Addressing review comments

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingContentFactory.java:
```diff
@@ -285,18 +295,24 @@ public ByteBuffer getBuffer()
-                FileTime lastModifiedTime = Files.getLastModifiedTime(_delegate.getResource().getPath());
-                if (lastModifiedTime.equals(_lastModifiedValue))
+                if (!_delegate.getResource().exists())
+                    return false;
```
Aren't these redundant checks now that the constructor checks them?

joakime replied:
Done

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CompressedContentFormat.java:
```diff
@@ -15,20 +15,10 @@
-
     public static final CompressedContentFormat GZIP = new CompressedContentFormat("gzip", ".gz");
     public static final CompressedContentFormat BR = new CompressedContentFormat("br", ".br");
     public static final CompressedContentFormat[] NONE = new CompressedContentFormat[0];
```
Why there is not one for `zip`? Tests are continuously creating the `zip` format when there could be a constant here.

joakime replied:
zip is an oddball one.
It's not a supported standard for Accept-Encoding / Content-Encoding

sbordet replied:
Fair enough.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EtagUtil.java:
```diff
@@ -0,0 +1,251 @@
+/**
+ * Utility classes for Etag behaviors
+ */
+public class EtagUtil
```
Contains more than one utility, so I'd prefer `EtagUtils`.

Also, add a `private` empty constructor so it cannot be instantiated.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EtagUtil.java:
```diff
@@ -0,0 +1,251 @@
+     * @param suffix the suffix for the ETag
+     * @return the weak etag
+     */
+    public static String calcWeakEtag(Path path, String suffix)
```
Prefix `calc` implies calculations, which to me implies math operations only.
I think something like `compute` or `resolve` instead of `calc` is more appropriate.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceContentFactory.java:
```diff
@@ -91,10 +91,14 @@ private HttpContent load(String pathInContext, Resource resource)
+                    continue; // skip, compressed resource modified before actual resource
+                if (compressedResource.length() >= resource.length())
+                    continue; // skip, not smaller than actual resource
+                // add found precompressed format
```
Remove the comments, they are just describing what the code does, and the code is very clear.
Less chances that the comments rot.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EtagUtil.java:
```diff
@@ -0,0 +1,251 @@
+        if (suffix != null)
+            b.append(suffix);
+        b.append('"');
+        return b.toString();
```
Is there a reference for this algorithm? Is it defined by some spec? If so, please add a reference to it.

joakime replied:
The Etag is just a string.
The bits and pieces we use are just our take on how to create that string.
I'll add the link to the RFC9110 section in the top level javadoc.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EtagUtil.java:
```diff
@@ -0,0 +1,251 @@
+     * @param suffix the suffix for the ETag
+     * @return the weak etag
+     */
+    public static String calcWeakEtag(Path path, String suffix)
```
The original implementation in `Resource` was not using `Path` at all, so I have the same concerns: this method should take a `Resource`, not a `Path`.

joakime replied:
The `Resource` version didn't handle non-path/file-system based Resources properly.

Btw, there is a `Resource` method in this class for working with `Resource` to `HttpField`.

joakime replied:
There's also the problem when working with a `ResourceCollection` there is a difference in `.getPath()` and `.getLastModified()` which doesn't match the Resource actually being served in the `HttpContent`

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EtagUtil.java:
```diff
@@ -0,0 +1,251 @@
+     * @return True if the tags are equal.
+     * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#section-8.8.3.2">RFC9110: Section 8.8.3.2 - Etag Comparison</a>.
+     */
+    public static boolean match(String etag, String etagWithOptionalSuffix)
```
I would call this method `matches` because it returns a boolean.
It's kind of a query/question rather than a command.

joakime committed:
Testing for whole content

joakime committed:
Rename EtagUtil to EtagUtils

joakime committed:
Improve MemoryResource handling of name/filename

joakime committed:
EtagUtils handling of Resource

joakime committed:
Checkstyle and compile fixes

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingContentFactory.java:
```diff
@@ -285,18 +295,24 @@ public ByteBuffer getBuffer()
                 {
-                    _lastAccessed = NanoTime.now();
-                    return true;
+                    FileTime lastModifiedTime = Files.getLastModifiedTime(_delegate.getResource().getPath());
```
Since this makes caching only work with path-backed resources, shouldn't there be a check in the constructor to prevent accepting those?

joakime replied:
Good point, let me fix this code block.

joakime committed:
Better protection of bad Resource impls in EtagUtils

joakime committed:
Update CachingContentFactory

+ Using Resource.lastModified were possible
+ CachingHttpContent has better Resource
  protection, and uses Etag from delegate
+ CachingHttpContent uses delegate where
  possible.
+ Store Etag HttpField in CachingHttpContent
  and allow override in constructor

joakime committed:
Simpler CachingHttpContent.isValid()

joakime committed:
Better handling of PrecompressedHttpContent
